### PR TITLE
STORM-3733 fix AsyncLocalizer from updating dead topology blobs

### PR DESCRIPTION
**Pull request by agresch**
## What is the purpose of the change

When a local assignment changes for a slot, the blob references are not updated. This causes a mismatch between the slot PNA and the blob reference PNA. When the topology is killed, the blob tries to remove the slot PNA and cannot find the reference. The AsyncLocalizer still has the old slot PNA as a reference, so cannot release the blob.

The local assignment could change due to a reordering of the same executors on a nimbus restart or something similar.

The fix is basically removeReference() changes in LocallyCachedBlob.  We make sure that we are checking for an equivalent PNA object instead of equals.  Additional logging was also added to track down this issue.  I think it's not excessive and considering how long this problem has existed, important to have.

## How was the change tested

Ran storm-server unit tests.  This change has been tested internally by running integration tests and on our staging clusters for a couple weeks and validated that we no longer get AsyncLocalizer alerts for this condition.

### CODE REVIEW

File: storm-server/src/main/java/org/apache/storm/utils/EquivalenceUtils.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            return false;
+        }
+        if (first == second) {
+            return true;
+        }
+        if (first.equals(second)) {
+            return true;
+        }
+
+        if (first.get_cpu() != second.get_cpu()) {
```
Comment (bipinprasad) [inline storm-server/src/main/java/org/apache/storm/utils/EquivalenceUtils.java:83]: Can second be null when first is not null? If so, an NPE here.

Comment (agresch) [reply storm-server/src/main/java/org/apache/storm/utils/EquivalenceUtils.java:86]: updated

File: storm-server/src/main/java/org/apache/storm/utils/EquivalenceUtils.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+     * This method compares WorkerResources while considering any resources are NULL to be 0.0
+     *
+     * @param first  WorkerResources A
+     * @param second WorkerResources B
+     * @return True if A and B are equivalent, treating the absent resources as 0.0
+     */
+    @VisibleForTesting
+    static boolean customWorkerResourcesEquality(WorkerResources first, WorkerResources second) {
+        if (first == null) {
```
Comment (Ethanlm) [inline storm-server/src/main/java/org/apache/storm/utils/EquivalenceUtils.java:73]: It is a little wired that `customWorkerResourcesEquality(null, null)` would return `false`. But since this is not related to this PR as the code is moved from other places in storm, I am not worried about it in this PR

**Review by Ethanlm — APPROVED** [review]
Looks good to me
